kong/core/plugins_iterator.lua

```diff
@@ -63,52 +118,13 @@ local function iter_plugins_for_req(loaded_plugins, access_or_cert_ctx)
     ctx.plugins_for_request = {}
   end
 
-  local i = 0
-
-  local function get_next()
-    i = i + 1
-    local plugin = loaded_plugins[i]
-    local api = ctx.api
-    if plugin then
-      -- load the plugin configuration in early phases
-      if access_or_cert_ctx then
-
-        local plugin_configuration
-
-        -- Search API and Consumer specific, or consumer specific
-        local consumer_id = (ctx.authenticated_consumer or empty).id        
-        if consumer_id and plugin.schema and not plugin.schema.no_consumer then
-          if api then
-            plugin_configuration = load_plugin_configuration(api.id, consumer_id, plugin.name)
-          end
-          if not plugin_configuration then
-            plugin_configuration = load_plugin_configuration(nil, consumer_id, plugin.name)
-          end
-        end
-
-        if not plugin_configuration then
-          -- Search API specific, or global
-          if api then
-            plugin_configuration = load_plugin_configuration(api.id, nil, plugin.name)
-          end
-          if not plugin_configuration then
-            plugin_configuration = load_plugin_configuration(nil, nil, plugin.name)
-          end
-        end
-
-        ctx.plugins_for_request[plugin.name] = plugin_configuration
-      end
-
-      -- return the plugin configuration
-      if ctx.plugins_for_request[plugin.name] then
-        return plugin, ctx.plugins_for_request[plugin.name]
-      end
-
-      return get_next() -- Load next plugin
-    end
-  end
+  plugin_iter_state.i                  = 0
+  plugin_iter_state.ctx                = ctx
+  plugin_iter_state.api                = ctx.api
+  plugin_iter_state.loaded_plugins     = loaded_plugins
+  plugin_iter_state.access_or_cert_ctx = access_or_cert_ctx
```

Comment:
Instead of creating a table for this, I'm wondering whether it would be more performant to pass these 5 values as parameters to `get_next`. Since the value returned by `iter_plugins_for_req` must be callable, we would have to return a function, so I don't know.

```
function iter_plugins_for_req(loaded_plugins, access_or_cert_crx)
  local ctx = ngx.ctx
  if not ctx.plugins_for_request then
    ctx.plugins_for_request = {}
  end

  return function() get_next(0, ctx, ctx.api, loaded_plugins, access_or_cert_ctx) end
end
```